internal/failure/failure.go

```diff
@@ -0,0 +1,268 @@
+	DefaultLine = 1
+	// DefaultColumn is the default Failure column number.
+	DefaultColumn = 1
+)
+
+var (
+	_defaultFields = []Field{
+		Filename,
+		Line,
+		Column,
+		Message,
+	}
+	_stringToField = map[string]Field{
+		"filename": Filename,
+		"line":     Line,
+		"column":   Column,
+		"id":       Identifier,
+		"message":  Message,
+	}
+	_IDToString = map[ID]string{
+		Error:  "ERROR",
+		Format: "FORMAT",
+		Lint:   "LINT",
+		Proto:  "PROTO",
+	}
+)
+
+// String implements fmt.Stringer.
+func (id ID) String() string {
+	if s, ok := _IDToString[id]; ok {
+		return s
+	}
+	return strconv.Itoa(int(id))
+}
+
+// ParseFields parses Fields from the given string.
+// Fields are expected to be colon-separated in the given string.
+// Input is case-insensitive. If the string is empty, _defaultFields
+// will be returned.
+func ParseFields(s string) ([]Field, error) {
+	if len(s) == 0 {
+		return _defaultFields, nil
+	}
+	fields := strings.Split(s, ":")
+	failureFields := make([]Field, len(fields))
+	for i, f := range fields {
+		ff, err := parseField(f)
+		if err != nil {
+			return nil, err
+		}
+		failureFields[i] = ff
+	}
+	return failureFields, nil
+}
+
+// Failure is a failure with a position in text.
+type Failure struct {
+	Filename string
+	Line     int
+	Column   int
+	ID       string
+	Message  string
+}
+
+// Writer is a writer that Failure.Println can accept.
+//
+// Both bytes.Buffer and bufio.Writer implement this.
+type Writer interface {
+	WriteRune(rune) (int, error)
+	WriteString(string) (int, error)
+}
+
+// Fprintln prints the Failure to the writer with the given ordered fields.
+// The given fields will overwrite the fields already set in this Failure.
+func (f *Failure) Fprintln(writer Writer, fields ...Field) error {
+	if len(fields) == 0 {
+		fields = _defaultFields
+	}
+	printColon := true
+	for i, field := range fields {
+		switch field {
+		case Filename:
+			filename := f.Filename
+			if filename == "" {
+				filename = DefaultFilename
+			}
+			if _, err := writer.WriteString(filename); err != nil {
+				return err
+			}
+		case Line:
+			line := strconv.Itoa(f.Line)
+			if line == "0" {
+				line = strconv.Itoa(DefaultLine)
+			}
+			if _, err := writer.WriteString(line); err != nil {
+				return err
+			}
+		case Column:
+			column := strconv.Itoa(f.Column)
+			if column == "0" {
+				column = strconv.Itoa(DefaultColumn)
+			}
+			if _, err := writer.WriteString(column); err != nil {
+				return err
+			}
+		case Identifier:
+			id := f.ID
+			if id == "" {
+				id = Error.String()
+			}
+			if _, err := writer.WriteString(id); err != nil {
+				return err
+			}
+			printColon = false
+		case Message:
+			if _, err := writer.WriteRune(' '); err != nil {
+				return err
+			}
+			if _, err := writer.WriteString(f.Message); err != nil {
+				return err
+			}
+			printColon = false
+		default:
+			return fmt.Errorf("unknown Field: %v", field)
+		}
+		if printColon && i != len(fields)-1 {
```

Comment:
By removing the `written` boolean, you now print a colon even if the field was empty. The `written` boolean should be re-introduced.

Reply:
As an example, here's what `prototool lint` prints out now:

```
$ prototool lint internal/cmd/testdata/lint/lots.proto
internal/cmd/testdata/lint/lots.proto:1:1: File option "go_package" is required.
internal/cmd/testdata/lint/lots.proto:1:1: File option "java_package" is required.
```

Note the extra space after `1:1:`. In the old version, this was:

```
$ prototool lint internal/cmd/testdata/lint/lots.proto
internal/cmd/testdata/lint/lots.proto:1:1:File option "go_package" is required.
internal/cmd/testdata/lint/lots.proto:1:1:File option "java_package" is required.
```